compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -88,6 +88,33 @@ impl<'tcx> UnsafetyVisitor<'_, 'tcx> {
         }
     }
 
+    fn emit_deprecated_safe_fn_call(&self, span: Span, kind: &UnsafeOpKind) -> bool {
+        match kind {
+            // Allow calls to deprecated-safe unsafe functions if the caller is
+            // from an edition before 2024.
+            &UnsafeOpKind::CallToUnsafeFunction(Some(id))
+                if !span.at_least_rust_2024()
+                    && self.tcx.has_attr(id, sym::rustc_deprecated_safe_2024) =>
```

Comment:
While this PR should probably go through as-is to get the fix in as soon as possible, I feel like this system is somewhat brittle, relying on specifically the symbol `rustc_deprecated_safe_2024` and the edition being hard coded to check at least `2024`.  I wonder if it would be a good idea for a follow up PR to somehow make this general in that the attribute could specify an edition in which a function starts becoming unsafe, and check that edition?

However, such a followup assumes that we will make more functions `unsafe` in  the future due to oversight.  Ideally we don't have to do that, and I think that the current state of things is such that just about everything has been audited?  So perhaps it's not worth such a generalization, since it ideally doesn't happen again.

Reply:
> While this PR should probably go through as-is to get the fix in as soon as possible, I feel like this system is somewhat brittle, relying on specifically the symbol `rustc_deprecated_safe_2024` and the edition being hard coded to check at least `2024`. I wonder if it would be a good idea for a follow up PR to somehow make this general in that the attribute could specify an edition in which a function starts becoming unsafe, and check that edition?

Such a more general has been proposed and can probably be implemented. `#[rustc_deprecated_safe_2024]` was intentionally limited in scope in order to make `std::env::{set_var, remove_var}` unsafe in Rust 2024.